lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js

```diff
@@ -0,0 +1,132 @@
+                let o = this;
+
+                if (!(o instanceof __chakraLibrary.ArrayIterator)) {
+                    throw new TypeError("Array Iterator.prototype.next: 'this' is not an Array Iterator object");
+                }
+
+                let a = o.__$arrayObj$__;
+                let value, done;
+
+                if (o.__$internalDone$__ === true) {
+                    value = undefined;
+                    done = true;
+                }
+                else {
+                    let index = o.__$nextIndex$__;
+                    let len = __chakraLibrary.isArray(a) ? a.length : __chakraLibrary.GetLength(a);
+
+                    if (index < len) { // < comparison should happen instead of >= , because len can be NaN
+                        let itemKind = o.__$kind$__;
+
+                        o.__$nextIndex$__ = index + 1;
+
+                        if (itemKind === 1 /*ArrayIterationKind.Value*/) {
+                            value = a[index];
+                        }
+                        else if (itemKind === 0 /*ArrayIterationKind.Key*/) { // TODO (megupta) : Use clean enums here ?
+                            value = index;
+                        }
+                        else {
+                            let elementKey = index;
+                            let elementValue = a[index];
+                            value = [elementKey, elementValue];
+                        }
+                        done = false;
+                    }
+                    else {
+                        o.__$internalDone$__ = true;
+                        value = undefined;
+                        done = true;
+                    }
+                }
+                return { value: value, done: done };
+            },
+            writable: true,
+            enumerable: false,
+            configurable: true
+        }
+    );
+
+    ObjectDefineProperty(__chakraLibrary.ArrayIterator.prototype, Symbol.toStringTag, { value: "Array Iterator", writable: false, enumerable: false, configurable: true });
+
+    ObjectDefineProperty(__chakraLibrary.ArrayIterator.prototype.next, 'length', { value: 0, writable: false, enumerable: false, configurable: true });
+
+    ObjectDefineProperty(__chakraLibrary.ArrayIterator.prototype.next, 'name', { value: "next", writable: false, enumerable: false, configurable: true });
+
+    platform.registerChakraLibraryFunction("CreateArrayIterator", function (arrayObj, iterationKind) {
+        "use strict";
+        return new __chakraLibrary.ArrayIterator(arrayObj, iterationKind);
+    });
+
+    platform.registerFunction(FunctionsEnum.ArrayKeys, function () {
+        "use strict";
+        if (this === null || this === undefined) {
+            throw new TypeError("Array.prototype.keys: 'this' is null or undefined");
+        }
+        let o = __chakraLibrary.Object(this);
+        return __chakraLibrary.CreateArrayIterator(o, 0 /* ArrayIterationKind.Key*/);
```

Comment:
>0 /* ArrayIterationKind.Key*/ [](start = 54, length = 29)

Can these magic constants be turned into named constants or and/or members of the Enum?

Reply:
I cannot declare these constants within the global function we use to set up the library. It should be stored on the chakraLib object. Its not neat, but I want to avoid storing constants on the chakraLib object.